request/src/main/java/com/vimeo/networking2/internal/VimeoApiClientImpl.kt

```diff
@@ -0,0 +1,835 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.vimeo.networking2.internal
+
+import com.vimeo.networking2.*
+import com.vimeo.networking2.common.Followable
+import com.vimeo.networking2.config.Configuration
+import com.vimeo.networking2.enums.CommentPrivacyType
+import com.vimeo.networking2.enums.ConnectedAppType
+import com.vimeo.networking2.enums.EmbedPrivacyType
+import com.vimeo.networking2.enums.ErrorCodeType
+import com.vimeo.networking2.enums.ViewPrivacyType
+import com.vimeo.networking2.params.BatchPublishToSocialMedia
+import com.vimeo.networking2.params.ModifyVideoInAlbumsSpecs
+import com.vimeo.networking2.params.ModifyVideosInAlbumSpecs
+import com.vimeo.networking2.params.SearchDateType
+import com.vimeo.networking2.params.SearchDurationType
+import com.vimeo.networking2.params.SearchFacetType
+import com.vimeo.networking2.params.SearchFilterType
+import com.vimeo.networking2.params.SearchSortDirectionType
+import com.vimeo.networking2.params.SearchSortType
+import okhttp3.CacheControl
+
+/**
+ * The internal implementation of the [VimeoApiClient]. Once configured, this client cannot be re-configured.
+ *
+ * @param vimeoService The service used to make requests to the Vimeo API.
+ * @param authenticator The authenticator used to obtain tokens which can be used to make requests.
+ * @param configuration The configuration used by this client instance.
+ * @param basicAuthHeader The basic auth header using the client ID and secret, used if the account store does not
+ * provide an authenticated account.
+ * @param localVimeoCallAdapter The adapter used to notify consumers of local errors.
+ */
+@Suppress("LargeClass")
+internal class VimeoApiClientImpl(
+    private val vimeoService: VimeoService,
+    private val authenticator: Authenticator,
+    private val configuration: Configuration,
+    private val basicAuthHeader: String,
+    private val localVimeoCallAdapter: LocalVimeoCallAdapter
+) : VimeoApiClient {
+
+    private val authHeader: String
+        get() {
+            val vimeoAccount = authenticator.currentAccount
+
+            return vimeoAccount?.accessToken?.let { "Bearer $it" } ?: basicAuthHeader
+        }
+
+    override fun createAlbum(
+        name: String,
+        albumPrivacy: AlbumPrivacy,
+        description: String?,
+        parameters: Map<String, Any>?,
+        callback: VimeoCallback<Album>
+    ): VimeoRequest {
+        val body = parameters.intoMutableMap()
+        body[ApiConstants.Parameters.PARAMETER_ALBUM_NAME] = name
```

Comment:
It's my opinion that the defensive client side code was redundant and better off left out. It complicates the client logic and runs the risk of being out of date with the actual API requirements. My position is that it is better to let the API fail the request if there is an issue with a parameter. Now certain errors we must fail for client side, specifically those where there is a type mismatch. 

Coincidentally, the parameter this comment is on is a good demonstration of the above logic. The check in the old `VimeoClient` that prevented an album being created with an empty name was unnecessary. The API accepts empty names in the create request and substitutes "Untitled" in its place. I'm not sure if it was originally an invalid parameter when the old `VimeoClient` code was written, but I think it is a better idea to let the API tell us of such an issue rather than make those determinations here as much as possible.

Reply:
@anthonycr I agree less client validation logic as a general directive, but I still think there is more that can be done here in cases that are simply going to fail (for example, an empty request uri). There is never a case where API requirements will accommodate that. 

Additionally, I was also pointing this out because unless the removal of client-side logical changes have been individually checked for their impact on the client app, we are vulnerable to new bugs and/or crashes when these edge cases occur. I'm not sure whether you've vetted each of these instances or if it is just a risk you are willing to take.